File: openmp-wg-2016/omp_daxpy.f90
program omp_daxpy

   integer,parameter :: N=1e7
   real*8,save  :: x(N),y(N),z(N)
   real*8 :: a = 5./3.

   call tick(it0,it1)
   call daxpy(N,a,x,y,z)
   call tock(it0,it1)

end program

subroutine daxpy(N,a,x,y,z)
   integer :: N
   real*8  :: a,x(N),y(N),z(N)

!$OMP PARALLEL default(none) shared(N,a,x,y,z)

!$OMP DO
   do i = 1,N
      x(i) = real(i)*real(i)
      y(i) = (i+1.)*(i-1.)
   end do
!OMP end do

!$OMP DO
   do i = 1,N
      z(i) = x(i)*a+y(i)
   end do
!OMP end do
!$OMP END PARALLEL

end subroutine

File: openmp-wg-2016/omp_daxpy.c
// daxpy.c
//
// Compute a daxpy serially.
//
//

#include "ticktock.h"

void daxpy(int n, double a, double* x, double* y, double* z) 
{
    #pragma omp parallel default(none) shared(n,x,y,a,z)
    {
    int i;
    #pragma omp for
    for (i=0; i<n; i++) {
        x[i] = (double)i*(double)i;
        y[i] = (i+1.)*(i-1.);
    }
    #pragma omp for
    for (i=0; i<n; i++) 
        z[i] += a * x[i] + y[i];
    }
}

int main() 
{
    int     n = 1e7;
    double* x = malloc(sizeof(double) * n);
    double* y = malloc(sizeof(double) * n);
    double* z = malloc(sizeof(double) * n);
	double  a = 5./3.;
    tick_tock tt;
    tick(&tt);
    daxpy(n,a,x,y,z);
    tock(&tt);
    free(x);
    free(y);
    free(z);
}
